### PR TITLE
tests: re-write flaky test in bash

### LINKED ISSUE #3547: fix flaky unit test
cylc/flow/tests/network/test_client.py:test_serial_request

Observed failing on Travis, GitHub actions and locally.
https://github.com/oliver-sanders/cylc/pull/17/checks?check_run_id=555480556#step:7:1187

Possibly related, some unittests are leaving behind files in cylc-run.

### PR DESCRIPTION
**Pull request by oliver-sanders**
Closes #3547 

This is a functional test which was written in the pytest framework.

The `CylcWorkflowTestCase` harness is not really suitable for use with functional tests as yet and is very fragile to internal changes in the scheduler.

**Requirements check-list**
- [x] I have read `CONTRIBUTING.md` and added my name as a Code Contributor.
- [x] Contains logically grouped changes (else tidy your branch by rebase).
- [x] Does not contain off-topic changes (use other PRs for other changes).
- [x] Appropriate tests are included (unit and/or functional).
- [x] No change log entry required (why? e.g. invisible to users).
- [x] No documentation update required.

### CODE REVIEW

**Comment by oliver-sanders** [comment]
Unfortunately I think the other `CylcWorkflowTestCase` tests are also flaky 😞, they may need similar treatment.

File: tests/protobuf/00-basic.t [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+from cylc.flow.network.client import SuiteRuntimeClient
+from cylc.flow.network.server import PB_METHOD_MAP
+client = SuiteRuntimeClient('${SUITE_NAME}')
+pb_msg = client.serial_request('pb_entire_workflow')
+pb_data = PB_METHOD_MAP['pb_entire_workflow']()
+pb_data.ParseFromString(pb_msg)
+if pb_data.workflow.id != '${USER}|${SUITE_NAME}':
+    raise ValueError(f'incorrect flow id: {pb_data.workflow.id}')
+"
+cylc stop --now --now "${SUITE_NAME}"
```
Comment (kinow) [inline tests/protobuf/00-basic.t:51]: Extra --now?

File: tests/protobuf/00-basic.t [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+run_ok "${TEST_NAME_BASE}" python3 -c "
+from cylc.flow.network.client import SuiteRuntimeClient
+from cylc.flow.network.server import PB_METHOD_MAP
+client = SuiteRuntimeClient('${SUITE_NAME}')
+pb_msg = client.serial_request('pb_entire_workflow')
+pb_data = PB_METHOD_MAP['pb_entire_workflow']()
+pb_data.ParseFromString(pb_msg)
+if pb_data.workflow.id != '${USER}|${SUITE_NAME}':
+    raise ValueError(f'incorrect flow id: {pb_data.workflow.id}')
+"
```
Comment (kinow) [inline tests/protobuf/00-basic.t:50]: Are we able to load an external py file? This way linter/ide/grep can find the code more easily.

File: tests/protobuf/00-basic.t (hunk 1 above)
Comment (oliver-sanders) [reply tests/protobuf/00-basic.t:51]: Nope, `--now` the option we love so much we use it twice.

File: tests/protobuf/00-basic.t (hunk 2 above)
Comment (oliver-sanders) [reply tests/protobuf/00-basic.t:50]: It is possible, but then we wouldn't be able to template in variables (e.g. `SUITE_NAME`), worth it for 8 lines?

Comment (kinow) [reply tests/protobuf/00-basic.t:50]: Ah, true. Leave as is then.

File: tests/protobuf/00-basic.t [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+#-------------------------------------------------------------------------------
+# Test restarting a simple suite with a task in a retrying state.
```
Comment (hjoliver) [inline tests/protobuf/00-basic.t:18]: This comment seems to have broken through from another dimension?

**Comment by hjoliver** [comment]
Just assigned @dwsutherland (author of the original) in case he has wildly different views on this change.

**Comment by oliver-sanders** [comment]
(whoops, I thought it was Bruno's test because the suite was called five)

**Comment by oliver-sanders** [comment]
We want to move away from using perl to run the functional tests, I'm not sure we want to move away from bash per-se.

When we have viable Python API it would make a lot more sense to write tests in Python.

**Comment by oliver-sanders** [comment]
Dammit, now another unites is failing: cylc/flow/tests/network/test_publisher.py

**Comment by dwsutherland** [comment]
> Dammit, now another unites is failing: cylc/flow/tests/network/test_publisher.py
> 
> Perhaps we need a more general solution to these unit tests.
> 
> Hmmm, I have an idea...

They were consistent in Travis (but I'd take this for stable-ish & fast functional tests in Actions)...
I found dealing with threads difficult..

**Comment by oliver-sanders** [comment]
I think the reason they were stable in Travis is that we were running tests in series there, whereas on actions they are running in parallel. I always run tests in parallel on my own machine so I've been dealing with this for a while.

**Comment by oliver-sanders** [comment]
(following up on a route to running the scheduler without having to re-implement the whole thing in test code, if it's a quick win I may put it in)

**Comment by oliver-sanders** [comment]
Looks like there might be a nice and fairly simple solution to a Python suite harness which can cut out all the gunk and bring parallel safety at the same time...

**Comment by oliver-sanders** [comment]
I'm going to propose #3616 as a solution to this.

**Comment by oliver-sanders** [comment]
I'm feeling good enough about the integration test work to close this as superseded by #3616

**Pull request closed without merging**